ldap_auth_provider.py

```diff
@@ -139,73 +174,102 @@ def check_password(self, user_id, password):
                 )
                 defer.returnValue(False)
 
-            # check if user with user_id exists
-            if (yield self.account_handler.check_user_exists(user_id)):
-                # exists, authentication complete
-                yield threads.deferToThread(conn.unbind)
-                defer.returnValue(True)
+            # Fetch user data from
+            # existing ldap connection
+            query = "({prop}={value})".format(
+                prop=self.ldap_attributes['uid'],
+                value=localpart
+            )
 
-            else:
-                # does not exist, fetch metadata for account creation from
-                # existing ldap connection
-                query = "({prop}={value})".format(
-                    prop=self.ldap_attributes['uid'],
-                    value=localpart
+            if self.ldap_mode == LDAPMode.SEARCH and self.ldap_filter:
+                query = "(&{filter}{user_filter})".format(
+                    filter=query,
+                    user_filter=self.ldap_filter
                 )
+            logger.debug(
+                "ldap search filter: %s",
+                query
+            )
 
-                if self.ldap_mode == LDAPMode.SEARCH and self.ldap_filter:
-                    query = "(&{filter}{user_filter})".format(
-                        filter=query,
-                        user_filter=self.ldap_filter
-                    )
-                logger.debug(
-                    "ldap registration filter: %s",
-                    query
-                )
+            yield threads.deferToThread(
+                conn.search,
+                search_base=self.ldap_base,
+                search_filter=query,
+                attributes=self.ldap_attributes.values()
+            )
 
-                yield threads.deferToThread(
-                    conn.search,
-                    search_base=self.ldap_base,
-                    search_filter=query,
-                    attributes=[
-                        self.ldap_attributes['name'],
-                        self.ldap_attributes['mail']
-                    ]
-                )
+            responses = [
+                response
+                for response
+                in conn.response
+                if response['type'] == 'searchResEntry'
+            ]
 
-                if len(conn.response) == 1:
-                    attrs = conn.response[0]['attributes']
+            if len(responses) == 1:
+                attrs = responses[0]['attributes']
+                try:
                     name = attrs[self.ldap_attributes['name']][0]
-                    try:
-                        mail = attrs[self.ldap_attributes['mail']][0]
-                    except KeyError:
-                        mail = None
-
-                    # create account
+                except:
+                    name = None
+                try:
+                    mail = attrs[self.ldap_attributes['mail']][0]
```

Comment:
what happens when user object has multiple mail attributes? only first one will be used?